Add ChefSpec Tests and Fix Foodcritic Issue

pwelch opened the pull request:
- Add and Update ChefSpec Tests for current cookbook logic
- Fix Foodcritic FC007 Error by adding build-essential dependency to metadata

pwelch committed:
Add ChefSpec Tests

Add ChefSpec Tests for current cookbook logic.

pwelch committed:
Fix Foodcritic FC007 Error

Add build-essential dependency to metadata to fix Foodcritic
issue:

FC007 - Ensure recipe dependencies are reflected in cookbook metadata

jtimberman commented:
I'll follow up next week. Short list:

Windows should be a version that is used for workstations not a server version.

Please no ruby 1.9 hashes. E.g., `{key: "value"}`

Use the zero runner, as this is intended to be used with a server.

Thank you for this.

jtimberman commented:
Also testing recipe inclusion is an implementation detail that I'm not sure is going to stay.

Build-essential is a dependency of homebrew, that's why it wasn't declared here. I don't care about that foodcritic finding :)

jtimberman commented on metadata.rb:
```diff
@@ -6,3 +6,4 @@
 version           '0.2.0'
 depends           'homebrew', '~> 1.11'
 depends           'chocolatey', '~> 0.2'
+depends           'build-essential'
```
This is a dependency of homebrew.

jtimberman commented on spec/recipes/default_spec.rb:
```diff
@@ -1,14 +1,30 @@
-  let(:chef_run) { ChefSpec::Runner.new.converge(described_recipe) }
+  context 'Mac OS X' do
+    let(:chef_run) do
+      ChefSpec::SoloRunner.new(platform: 'mac_os_x', version: '10.10').
```
Please use ZeroRunner. I have no intention of supporting Chef Solo (directly) and would rather testing is done with Zero.

jtimberman commented on spec/recipes/default_spec.rb:
```diff
@@ -1,14 +1,30 @@
+    end
+
+    it 'includes pantry::mac_os_x recipe' do
+      expect(chef_run).to include_recipe('pantry::mac_os_x')
```
this include_recipe is ok

jtimberman commented on spec/recipes/default_spec.rb:
```diff
@@ -1,14 +1,30 @@
+    end
+
+    it 'includes pantry::windows recipe' do
+      expect(chef_run).to include_recipe('pantry::windows')
```
as is this one

jtimberman commented on spec/recipes/mac_os_x_spec.rb:
```diff
@@ -0,0 +1,34 @@
+    homebrew::cask
+    homebrew::install_formulas
+    homebrew::install_casks
+  }.each do |recipe|
```
however the current include_recipe for these is an implementation detail. I don't think it should be tested here.

pwelch replied:
Removed on 6a421399

jtimberman commented on spec/recipes/windows_spec.rb:
```diff
@@ -0,0 +1,12 @@
+  end
+
+  it 'includes chocolatey recipe' do
+    expect(chef_run).to include_recipe('chocolatey')
```
again, implementation detail - i don't think this should be tested specifically.

This like the other include_recipe statements in the os x specific recipe above are implementation details. There's little value here because they're unconditionally included, and then we're just testing that chef works. The include recipe for the platforms in general (in the default recipe) has value because we're ensuring that we get the correct thing based on the platform.

pwelch replied:
Removed on 4dd9f04

jtimberman commented on spec/spec_helper.rb:
```diff
@@ -2,6 +2,6 @@
 
 RSpec.configure do |config|
-  config.color_enabled = true
+  config.color     = true
```
Was this changed in rspec 3? For pete's sake.

pwelch replied:
As far as I can tell. When I run it with `color_enabled` it raises an error:
```shell
 pantry/spec/spec_helper.rb:5:in `block in <top (required)>': undefined method `color_enabled=' for #<RSpec::Core::Configuration:0x007ff5038299c0> (NoMethodError)
	from /Users/pwelch/.rvm/gems/ruby-2.1.5/gems/rspec-core-3.1.7/lib/rspec/core.rb:81:in `configure'
	from /Users/pwelch/Code/github/opensource/pantry/spec/spec_helper.rb:4:in `<top (required)>'
	from /Users/pwelch/Code/github/opensource/pantry/spec/recipes/default_spec.rb:1:in `require'
	from /Users/pwelch/Code/github/opensource/pantry/spec/recipes/default_spec.rb:1:in `<top (required)>'
	from /Users/pwelch/.rvm/gems/ruby-2.1.5/gems/rspec-core-3.1.7/lib/rspec/core/configuration.rb:1105:in `load'
	from /Users/pwelch/.rvm/gems/ruby-2.1.5/gems/rspec-core-3.1.7/lib/rspec/core/configuration.rb:1105:in `block in load_spec_files'
	from /Users/pwelch/.rvm/gems/ruby-2.1.5/gems/rspec-core-3.1.7/lib/rspec/core/configuration.rb:1105:in `each'
	from /Users/pwelch/.rvm/gems/ruby-2.1.5/gems/rspec-core-3.1.7/lib/rspec/core/configuration.rb:1105:in `load_spec_files'
	from /Users/pwelch/.rvm/gems/ruby-2.1.5/gems/rspec-core-3.1.7/lib/rspec/core/runner.rb:96:in `setup'
	from /Users/pwelch/.rvm/gems/ruby-2.1.5/gems/rspec-core-3.1.7/lib/rspec/core/runner.rb:84:in `run'
	from /Users/pwelch/.rvm/gems/ruby-2.1.5/gems/rspec-core-3.1.7/lib/rspec/core/runner.rb:69:in `run'
	from /Users/pwelch/.rvm/gems/ruby-2.1.5/gems/rspec-core-3.1.7/lib/rspec/core/runner.rb:37:in `invoke'
	from /Users/pwelch/.rvm/gems/ruby-2.1.5/gems/rspec-core-3.1.7/exe/rspec:4:in `<top (required)>'
	from /Users/pwelch/.rvm/gems/ruby-2.1.5/bin/rspec:23:in `load'
	from /Users/pwelch/.rvm/gems/ruby-2.1.5/bin/rspec:23:in `<main>'
	from /Users/pwelch/.rvm/gems/ruby-2.1.5/bin/ruby_executable_hooks:15:in `eval'
	from /Users/pwelch/.rvm/gems/ruby-2.1.5/bin/ruby_executable_hooks:15:in `<main>'
```
Setting to color seems to make Rspec 3 happy.

pwelch commented:
@jtimberman, Thanks for taking the time to review this! I'll get this fixed and pushed up.

pwelch committed:
Clean Up Spec Tests

- Remove build-essential as dependency. It is actually a dependency
of Homebrew cookbook and Foodcritic ignore
- Remove include_recipes for mac_os_x because it is an implemenation
detail that does not need to be tested
- Remove include_recipes for windows because it is an implementation
detail that does not need to be tested
- Remove windows spec tests because it is empty after removing
implementation tests
- Use ChefSpec ServerRunner instead of ChefSolo Runner

pwelch commented on spec/recipes/default_spec.rb:
```diff
@@ -1,14 +1,30 @@
-    pending 'Replace this with meaningful tests'
+  context 'Windows' do
+    let(:chef_run) do
+      ChefSpec::ServerRunner.new(:platform => 'windows', :version => '2012').
```
Not sure if I missed it but I only see Windows Server in Fauxhai [here](https://github.com/customink/fauxhai/tree/master/lib/fauxhai/platforms/windows). Is leaving this version okay?

jtimberman replied:
We should get Windows 8.1 ohai output added to fauxhai for sure. This is probably sufficient - I don't really plan on doing version specific support for Windows. The most relevant thing is that powershell of a sufficient version to run the pantry.ps1 is already installed.

jtimberman commented:
I'll look at this again in detail next week

jtimberman commented:
Haven't had cycles this week, sorry. Maybe next! :cake:

josegonzalez commented:
@jtimberman Would something like [tailor](https://github.com/turboladen/tailor) be of use to you in a PR? I'm using it at work for my cookbooks and it catches a few different stylistic issues (though I don't think it would catch the 1.9 bashes).

jtimberman commented:
I prefer Rubocop. It's included in ChefDK, and I use a specific `.rubocop.yml` in this repository.
